ethcore/src/snapshot/service.rs

```diff
@@ -327,9 +328,74 @@ impl Service {
 
 	// replace one the client's database with our own.
 	fn replace_client_db(&self) -> Result<(), Error> {
-		let our_db = self.restoration_db();
+		let rest_db = self.restoration_db();
+
+		let cur_chain = &self.client.chain();
+		let next_db = self.restoration_db_handler.open(&rest_db)?;
+		let next_chain = BlockChain::new(Default::default(), &[], next_db.clone());
+		let next_chain_info = next_chain.chain_info();
+
+		// Try to include every block that will need to be downloaded from the current chain
+		// Break when no more blocks are available from it.
+		match (next_chain_info.ancient_block_number, next_chain_info.first_block_number) {
+			(Some(next_ancient_block), Some(next_first_block)) if next_ancient_block + 1 < next_first_block => {
+				trace!(target: "snapshot", "Trying to import ancient blocks from {} to {}",
+					next_ancient_block, next_first_block,
+				);
+
+				let mut block_number = next_ancient_block + 1;
+				let mut batch = DBTransaction::new();
+				let mut parent_diff = None;
+
+				while block_number < next_first_block {
+					let chain = cur_chain.read();
+
+					if let Some(block_hash) = chain.block_hash(block_number) {
```

Comment:
we don't have an exclusive lock on the chain's DB. this is racing with client's normal block import. (in practice this isn't a problem right now but let's not lean on assumptions from outside the module)

Reply:
> The blocks won't change during it
why?
this is an assumption beyond the module.

Reply:
This is easiest when we import blocks in reverse because the `parent_hash` is always available. But then we should always make sure that `next_first_block.parent_hash` corresponds to the hash of a block that we import as its parent.

Reply:
So could it be just a check that for each block, the parent's hash is the one expected, ie. the one we got from the previous block ? If there is a mismatch, then just stop right there.